### PR TITLE
fix: avoid crash from channelPointRewardAdded callback

### LINKED ISSUE #2432: Chatterino SIGABRT due to redemptions
**Describe the bug**
Chatterino quits (it sends a SIGABRT signal) at apparently random times. It seems to be triggered by addChannelPointReward according to the backtrace.

**To reproduce**
I have no consistent way to reproduce the issue. It doesn't happen very often (5 times in the last 2 weeks). The bug triggers with no gui interaction, even when I'm on a different program's window.

**Screenshots**
Backtrace:
image
image

**Chatterino version**
Chatterino 2.2.3-beta2 (commit a6e23e246)

**Operating system**
Archlinux 5.10.11-arch1-1  qt 5.15.2

**Additional information**
My PC was only running on 4GB of RAM when all these crashes occurred, so maybe this had something to do. Now I have upgraded to 8GB, will report if the issue comes up again.

Full backtrace: https://pastebin.com/yFDuAGeC

### LINKED ISSUE #3942: Channel Point crash
I had a crash happen earlier when a channel point reward was added through pubsub.

https://github.com/Chatterino/chatterino2/blob/46f43f3ce82fbf6be1f4035d75d5ee8d6685f08d/src/providers/twitch/TwitchChannel.cpp#L254 fires and is consumed here https://github.com/Chatterino/chatterino2/blob/46f43f3ce82fbf6be1f4035d75d5ee8d6685f08d/src/providers/twitch/IrcMessageHandler.cpp#L383 but the connection is never disconnected, nor does it contain any sort of shared pointer logic, so the function can go out of scope causing a segmentation fault

backtrace:
```
(gdb) bt
#0  0x00007f827d8a14dc in  () at /usr/lib/libc.so.6
#1  0x00007f827d851998 in raise () at /usr/lib/libc.so.6
#2  0x00007f827d83b53d in abort () at /usr/lib/libc.so.6
#3  0x00007f827da99833 in __gnu_cxx::__verbose_terminate_handler() () at /usr/src/debug/gcc/libstdc++-v3/libsupc++/vterminate.cc:95
#4  0x00007f827daa5cfc in __cxxabiv1::__terminate(void (*)()) (handler=<optimized out>) at /usr/src/debug/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:48
#5  0x00007f827daa4cba in __cxa_call_terminate(_Unwind_Exception*) (ue_header=0x55aa14094c10) at /usr/src/debug/gcc/libstdc++-v3/libsupc++/eh_call.cc:54
#6  0x00007f827daa543a in __cxxabiv1::__gxx_personality_v0(int, _Unwind_Action, _Unwind_Exception_Class, _Unwind_Exception*, _Unwind_Context*)
    (version=<optimized out>, actions=6, exception_class=5138137972254386944, ue_header=<optimized out>, context=0x7ffd1adb9d10)
    at /usr/src/debug/gcc/libstdc++-v3/libsupc++/eh_personality.cc:688
#7  0x00007f827f286774 in _Unwind_RaiseException_Phase2 (exc=0x55aa14094c10, context=0x7ffd1adb9d10, frames_p=0x7ffd1adb9c18) at /usr/src/debug/gcc/libgcc/unwind.inc:64
#8  0x00007f827f2871ed in _Unwind_Resume (exc=0x55aa14094c10) at /usr/src/debug/gcc/libgcc/unwind.inc:242
#9  0x000055a9fb2d581e in chatterino::TwitchChannel::addChannelPointReward(chatterino::ChannelPointReward const&) (this=0x55a9ffc25610, reward=...)
    at /home/pajlada/git/chatterino2/src/providers/twitch/TwitchChannel.cpp:254
#10 0x000055a9faf9c889 in operator()() const (__closure=0x7f8210065bc8) at /home/pajlada/git/chatterino2/src/Application.cpp:498
#11 0x000055a9faf9ed12 in Event::~Event() (this=0x7f8210065bb0, __in_chrg=<optimized out>) at /home/pajlada/git/chatterino2/src/util/PostToThread.hpp:53
#12 0x000055a9faf9ed4a in Event::~Event() (this=0x7f8210065bb0, __in_chrg=<optimized out>) at /home/pajlada/git/chatterino2/src/util/PostToThread.hpp:54
#13 0x00007f827e88e48d in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) () at /usr/lib/libQt5Core.so.5
#14 0x00007f827e8d4478 in  () at /usr/lib/libQt5Core.so.5
#15 0x00007f827e289c6b in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#16 0x00007f827e2e0001 in  () at /usr/lib/libglib-2.0.so.0
#17 0x00007f827e287392 in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#18 0x00007f827e8d825c in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Core.so.5
#19 0x00007f827e88611c in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Core.so.5
#20 0x00007f827e890c49 in QCoreApplication::exec() () at /usr/lib/libQt5Core.so.5
#21 0x000055a9faf9a252 in chatterino::Application::run(QApplication&) (this=0x7ffd1adba3a0, qtApp=...) at /home/pajlada/git/chatterino2/src/Application.cpp:182
#22 0x000055a9fafe62d9 in chatterino::runGui(QApplication&, chatterino::Paths&, chatterino::Settings&) (a=..., paths=..., settings=...)
    at /home/pajlada/git/chatterino2/src/RunGui.cpp:245
#23 0x000055a9faf852b1 in main(int, char**) (argc=1, argv=0x7ffd1adbfdd8) at /home/pajlada/git/chatterino2/src/main.cpp:88
```

Chatterino commit: `46f43f3ce82fbf6be1f4035d75d5ee8d6685f08d`

### PR DESCRIPTION
**Pull request by iProdigy**
Closes #2432 
Closes #3942 

# Problem

It is possible for `channelPointRewardAdded` connections to be created that have a long lifetime (take this as a given, I explain how this happens in the "Follow-up" section). Upon callback lambda invocation, the captured variables will be released. However, if the invocation is unrelated to the desired `rewardId`, releasing the captured variables will cause a seg fault when a matching invocation occurs later on. (Technically this matching invocation will take the form of empty reward ids because of the earlier variable destruction)

# Solution

Clone objects used in lambda scope to avoid early destruction (also ensure `reward` is properly copied for each signal connection)

# Follow-up

While this PR avoids a common crash, there are remaining edge cases to be tackled in future PRs.

1. ~~It is possible that `channelPointRewardAdded` is invoked before a signal connection is made in `IrcMessageHandler` (when `isChannelPointRewardKnown` yields false). This is the scenario I described [before](https://github.com/Chatterino/chatterino2/issues/3942#issuecomment-1248397260). When this happens, now the first redemption associated with the reward will never be posted to chat (and the cloned objects are not freed). To solve this, there should be a mutex that prevents `channelPointRewardAdded.invoke` when we are in the `!channel->isChannelPointRewardKnown(rewardId)` block (and after acquiring the mutex, we should check `isChannelPointRewardKnown` once more)~~ Scratch this, I assumed multiple threads were at play

2. It is possible that the pubsub socket is not connected when the text redemption occurs. As a result, a signal connection will be made, but `invoke` would not be called. Again, the cloned objects would not be freed (until the next redemption for the same reward occurs, while pubsub is connected). So, we ought to implement a timeout where if x seconds pass without signal invocation, we release the message regardless.

### CODE REVIEW

**Comment by iProdigy** [comment]
I just received `chatterino.twitch: TwitchChannel reward added ADD callback since reward is not known: ""` (not sure which channel), so bad payloads from twitch are not out of the question (in contributing to the edge cases that trigger the crashes)

**Comment by Nerixyz** [comment]
Maybe I'm reading this wrong, but I think the use of `SelfDisconnectingSignal` is wrong here, since it uses [`std::remove_if`](https://en.cppreference.com/w/cpp/algorithm/remove) which says that calling the callback shouldn't modify the value, which we kind-of do here by deleting the clone in the callback.

File: src/providers/twitch/IrcMessageHandler.cpp [hunk 1]
```diff
@@ -1251,23 +1251,31 @@ void IrcMessageHandler::addMessage(Communi::IrcMessage *message,
...
     {
         const auto rewardId = it.value().toString();
-        if (!channel->isChannelPointRewardKnown(rewardId))
+        if (!rewardId.isEmpty() &&
+            !channel->isChannelPointRewardKnown(rewardId))
         {
             // Need to wait for pubsub reward notification
+            auto *rewardClone = new QString(rewardId);
+            auto *targetClone = new QString(target);
+            auto *contentClone = new QString(originalContent);
```
[the remark this answers is not in the document]
Comment (iProdigy) [reply src/providers/twitch/IrcMessageHandler.cpp:1260]: Are you sure? Here's some logs from a recent crash I captured:

```
chatterino.twitch: [TwitchChannel "removed" ] Channel point reward added: "b37475ab-5bb3-4787-bdf9-d8c3582b62f5" , "Ask me anything" , true
chatterino.twitch: TwitchChannel reward added callback: "b37475ab-5bb3-4787-bdf9-d8c3582b62f5" - ""
chatterino.twitch: TwitchChannel reward added callback: "" -
Segmentation fault (core dumped)
```

Notice that `rewardId` has become an empty QString (and seemingly isn't even properly printed in the last callback log)

**Comment by iProdigy** [comment]
Working on full refactor à la https://github.com/Chatterino/chatterino2/pull/4943#pullrequestreview-1713956782

**Pull request closed without merging**